<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.12: http://docutils.sourceforge.net/" />
<title>Reactant/Product Pair Prediction Using FindPrimaryPairs</title>
<style type="text/css">

/*
:Author: David Goodger (goodger@python.org)
:Id: $Id: html4css1.css 7614 2013-02-21 15:55:51Z milde $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.

See http://docutils.sf.net/docs/howto/html-stylesheets.html for how to
customize this style sheet.
*/

/* used to remove borders from tables and images */
.borderless, table.borderless td, table.borderless th {
  border: 0 }

table.borderless td, table.borderless th {
  /* Override padding for "table.docutils td" with "! important".
     The right padding separates the table cells. */
  padding: 0 0.5em 0 0 ! important }

.first {
  /* Override more specific margin styles with "! important". */
  margin-top: 0 ! important }

.last, .with-subtitle {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

a.toc-backref {
  text-decoration: none ;
  color: black }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

object[type="image/svg+xml"], object[type="application/x-shockwave-flash"] {
  overflow: hidden;
}

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title, .code .error {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em ;
  margin-right: 2em }

div.footer, div.header {
  clear: both;
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin: 0 0 0.5em 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.section-subtitle, h2.section-subtitle, h3.section-subtitle,
h4.section-subtitle, h5.section-subtitle, h6.section-subtitle {
  margin-top: 0.4em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

img.align-left, .figure.align-left, object.align-left {
  clear: left ;
  float: left ;
  margin-right: 1em }

img.align-right, .figure.align-right, object.align-right {
  clear: right ;
  float: right ;
  margin-left: 1em }

img.align-center, .figure.align-center, object.align-center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.align-left {
  text-align: left }

.align-center {
  clear: both ;
  text-align: center }

.align-right {
  text-align: right }

/* reset inner alignment in figures */
div.align-right {
  text-align: inherit }

/* div.align-center * { */
/*   text-align: left } */

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font: inherit }

pre.literal-block, pre.doctest-block, pre.math, pre.code {
  margin-left: 2em ;
  margin-right: 2em }

pre.code .ln { color: grey; } /* line numbers */
pre.code, code { background-color: #eeeeee }
pre.code .comment, code .comment { color: #5C6576 }
pre.code .keyword, code .keyword { color: #3B0D06; font-weight: bold }
pre.code .literal.string, code .literal.string { color: #0C5404 }
pre.code .name.builtin, code .name.builtin { color: #352B84 }
pre.code .deleted, code .deleted { background-color: #DEB0A1}
pre.code .inserted, code .inserted { background-color: #A3D289}

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

span.section-subtitle {
  /* font-size relative to parent (h1..h6 element) */
  font-size: 80% }

table.citation {
  border-left: solid 1px gray;
  margin-left: 1px }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid 1px black;
  margin-left: 1px }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

table.docutils th.field-name, table.docinfo th.docinfo-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap ;
  padding-left: 0 }

/* "booktabs" style (no vertical lines) */
table.docutils.booktabs {
  border: 0px;
  border-top: 2px solid;
  border-bottom: 2px solid;
  border-collapse: collapse;
}
table.docutils.booktabs * {
  border: 0px;
}
table.docutils.booktabs th {
  border-bottom: thin solid;
  text-align: left;
}

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

ul.auto-toc {
  list-style-type: none }

</style>
</head>
<body>
<div class="document" id="reactant-product-pair-prediction-using-findprimarypairs">
<h1 class="title">Reactant/Product Pair Prediction Using FindPrimaryPairs</h1>

<p>This tutorial will go over how to use the <tt class="docutils literal">primarypairs</tt> function in <cite>PSAMM</cite>
to predict reactant/product pairs in genome scale models.</p>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><a class="reference internal" href="#materials" id="id3">Materials</a></li>
<li><a class="reference internal" href="#reactant-product-pair-prediction-using-psamm" id="id4">Reactant/Product Pair Prediction using <cite>PSAMM</cite></a></li>
<li><a class="reference internal" href="#visualizating-models-using-psamm" id="id5">Visualizating Models using PSAMM</a></li>
</ul>
</div>
<div class="section" id="materials">
<h1><a class="toc-backref" href="#id3">Materials</a></h1>
<p>For information on how to install <cite>PSAMM</cite> and the associated requirements, as well
how to download the materials required for this tutorial you can reference the
Installation and Materials section of the tutorial.</p>
<p>In addition to the basic installation of <cite>PSAMM</cite>, the visualization function can use
the <cite>Graphviz</cite> program to generate images from the text based graph format that will
be generated by the <tt class="docutils literal">vis</tt> command. <cite>Graphviz</cite> version &gt; 0.8.4 must be installed,
in addition the <cite>Graphviz</cite> python bindings must be installed in the working python
environment.</p>
<div class="note">
<p class="first admonition-title">Note</p>
<p>Graphviz download: <a class="reference external" href="https://www.graphviz.org/download/">https://www.graphviz.org/download/</a></p>
<p class="last">Graphviz python bindings: <a class="reference external" href="https://pypi.org/project/graphviz/">https://pypi.org/project/graphviz/</a>
or
(psamm-env) $ pip install graphviz</p>
</div>
<p>For this part of the tutorial we will be using a modified version of the E. coli
core metabolic models that has been used in the other sections of the tutorial.
This model has been modified to add in a new pathways for the utilization of
mannitol as a carbon source. To access this model and the other files needed you
will need to go into the tutorial-part-4 folder located in the psamm-tutorial folder.</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span><span class="name builtin">cd</span> &lt;PATH&gt;/tutorial-part-4/
</pre>
<p>Once in this folder you should see a folder called E_coli_yaml. inside of this folder
is all of the data that will be needed to run the commands in this tutorial. Inside of
this folder there will be a directory called E_coli_yaml/ that contains the E. coli
core model, and a directory called additional_files/ that contains some additional
input files that will be used in the visualization portion of the tutorial.</p>
<p>To run the following tutorials go into the E_coli_yaml/ directory:</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span><span class="name builtin">cd </span>E_coli_yaml/
</pre>
</div>
<div class="section" id="reactant-product-pair-prediction-using-psamm">
<h1><a class="toc-backref" href="#id4">Reactant/Product Pair Prediction using <cite>PSAMM</cite></a></h1>
<p>Metabolism can be broken down into individual metabolic reactions which transfer elements
between different metabolites. When looking at individual metabolic reactions it may be
possible to derive the element transfer patterns just based on previous knowledge of
the enzyme and biochemistry. Take the following reaction as an example:</p>
<pre class="code shell literal-block">
Acetate + ATP &lt;<span class="operator">=</span>&gt; Acetyl-Phosphate + ADP
</pre>
<p>This reaction is catalyzed by the enzyme Acetate Kinase. In this reaction acetate is
converted to acetyl-phosphate through the addition of a phosphate group from ATP.
A basic understanding of phosphorylation and the biological role of ATP makes
it possible to manually predict that the primary element transfers for
non hydrogen elements are as follows:</p>
<div class="system-message">
<p class="system-message-title">System Message: ERROR/3 (<tt class="docutils">fpp.rst</tt>, line 77)</p>
<p>Malformed table.
Text in column margin in table line 4.</p>
<pre class="literal-block">
=====================         ================
Reactant/Product Pair         Element Transfer
=====================         ================
Acetate -&gt; Acetyl-Phosphate   carbon backbone
ATP -&gt; ADP                    carbon backbone
ATP -&gt; Acetyl-Phosphate       phosphate group
Acetate -&gt; ADP                None
=====================         ================
</pre>
</div>
<p>While doing this for one or two simple reactions is possible, genome scale models often
contain hundreds to thousands of reactions, making manual reactant/product pair prediction
impractical. In addition to this, reaction mechanisms are often not known, and patterns
of element transfer within reactions are not available for most metabolic reactions.</p>
<p>To address this problem the <cite>FindPrimaryPairs</cite> algorithm [CITATION for FPP PAPER] was
developed and implemented within the new <cite>PSAMM</cite> function <tt class="docutils literal">primarypairs</tt>.</p>
<p>The <cite>FindPrimaryPairs</cite> is an iterative algorithm which can be used to predict element
transferring reactant/product pairs in genome scale models. <cite>FindPrimaryPairs</cite> relies
on two sources of information which are generally available in genome scale models,
reaction stoichiometry and metabolite formulas. From this information <cite>FindPrimaryPairs</cite>
can make a global prediction of element transferring reactant/product pairs without any
additional information about reaction mechanisms.</p>
<p>This kind of reactant/product pair prediction [[[importance statement]]]</p>
<div class="section" id="basic-use-of-the-primarypairs-command">
<span id="exclude-fpp"></span><h2>Basic Use of the <tt class="docutils literal">primarypairs</tt> Command</h2>
<p>The <tt class="docutils literal">primarypairs</tt> command in PSAMM can be used to perform an element transfering pair
prediction using the <cite>FindPrimaryPairs</cite> algorithm. The basic command can be run like so:</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span>psamm-model primarypairs --exclude &#64;./exclude.tsv
</pre>
<p>This function often requires a file to be provided through the <tt class="docutils literal"><span class="pre">--exclude</span></tt> option. This file
is a single column list of reaction IDs, of any reactions that the user wants to remove from the
model when doing the reactant/product pair prediction. Typically this file should contain any
artificial reactions that might be in the model such as Biomass objective reactions, macromolecule
synthesis reactions, etc. While these reactions can be left in the model, the fractional stoichiometries
and presence of artificial metabolites in the reaction can cause the algorithm to take a much longer
time to find a solution. In the case of this example with the E. coli core model the only reaction
like this is the biomass reaction <tt class="docutils literal">Biomass_Ecoli_core_w_GAM</tt>. And this is the only reaction listed
in the <cite>exlcude.tsv</cite> file.</p>
<div class="note">
<p class="first admonition-title">Note</p>
<p class="last">The <cite>FindPrimaryPairs</cite> algorithm relies on metabolite formulas to make its reactant/product pair
predictions. If any reactions contain metabolites that do not have any they will automatically be
excluded from the primary pair prediction.</p>
</div>
<p>When the above command is run the output will look like the following:</p>
<pre class="code shell literal-block">
INFO: Model: Ecoli_core_model
INFO: Model version: 3ac8db4
INFO: Using default element weights <span class="keyword">for</span> fpp: <span class="name variable">C</span><span class="operator">=</span>1, <span class="name variable">H</span><span class="operator">=</span>0, *<span class="operator">=</span>0.82
INFO: Iteration 1: <span class="literal number">79</span> reactions...
INFO: Iteration 2: <span class="literal number">79</span> reactions...
INFO: Iteration 3: <span class="literal number">8</span> reactions...
GLNS    nh4_c<span class="operator">[</span>c<span class="operator">]</span>        h_c<span class="operator">[</span>c<span class="operator">]</span>  H
FBA     fdp_c<span class="operator">[</span>c<span class="operator">]</span>        g3p_c<span class="operator">[</span>c<span class="operator">]</span>        C3H5O6P
ME2     mal_L_c<span class="operator">[</span>c<span class="operator">]</span>      nadph_c<span class="operator">[</span>c<span class="operator">]</span>      H
MANNI1PDEH      manni1p<span class="operator">[</span>c<span class="operator">]</span>      nadh_c<span class="operator">[</span>c<span class="operator">]</span>       H
PTAr    accoa_c<span class="operator">[</span>c<span class="operator">]</span>      coa_c<span class="operator">[</span>c<span class="operator">]</span>        C21H32N7O16P3S
....
</pre>
<p>In the first few lines basic information about the model name and version are provided. In the next
line the element weights used in the <cite>FindPrimaryPairs</cite> algorithm are listed (more on this below).
Then, as the algorithm goes through multiple iterations it will print out the iteration number and
how many reactions it is still trying to make a pair prediction for. After this information a four
column table is then printed out that contains the following columns from left to right, Reaction ID,
reactant ID, product ID, and elements transferred.</p>
<p>From this output the Acetate Kinase reaction from the above example can be looked at and compared to
the manual prediction of the element transfer. The reaction ID for this reaction is ACKr:</p>
<pre class="code shell literal-block">
ACKr    atp_c<span class="operator">[</span>c<span class="operator">]</span>        adp_c<span class="operator">[</span>c<span class="operator">]</span>        C10H12N5O10P2
ACKr    atp_c<span class="operator">[</span>c<span class="operator">]</span>        actp_c<span class="operator">[</span>c<span class="operator">]</span>       O3P
ACKr    ac_c<span class="operator">[</span>c<span class="operator">]</span> actp_c<span class="operator">[</span>c<span class="operator">]</span>       C2H3O2
</pre>
<p>From this result it can be seen that the prediction contains the same three element transferring pairs
as the above manual prediction; ATP -&gt; ADP, ATP -&gt; Acetyl-Phosphate, Acetate to Acetyl-Phosphate.</p>
<p>This basic usage of the <tt class="docutils literal">primarypairs</tt> command will allow for quick accurate prediction of element
transferring pairs in any of the reactions in a genome scale model, but the function also has a few
additional options that can be used to refine and adjust how the pair prediction work and what it will
produce as a result.</p>
</div>
<div class="section" id="modifying-element-weights">
<h2>Modifying Element Weights</h2>
<p>The metabolite pair prediction relies on a parameter called element weight to inform the algorithms
about what chemical elements should be considered more or less important when determine metabolite
similarity. An example of how this might be used can be seen in the default element weights that are
reported when running <tt class="docutils literal">primarypairs</tt>.</p>
<pre class="code shell literal-block">
INFO: Using default element weights <span class="keyword">for</span> fpp: <span class="name variable">C</span><span class="operator">=</span>1, <span class="name variable">H</span><span class="operator">=</span>0, *<span class="operator">=</span>0.82
</pre>
<p>These element weights are the default weights used when running <tt class="docutils literal">primarypairs</tt> with the <cite>FindPrimaryPairs</cite>
algorithm. In this case a weight of 1 is given to carbon. Because carbon forms the structural backbone of many
metabolites this element is given the most weight. In contrast, hydrogen is not usually a major structural
element within metabolites. This leads to a weight of 0 being given to hydrogen, meaning that it is not considered
when comparing formulas between two metabolites. By default all other elements are given an intermediate weight
of 0.82.</p>
<p>These default element weights can be adjusted using the <cite>--weights</cite> command line argument. For example, to adjust
the weight of the element nitrogen while keeping the other elements the same as the default settings the following
command could be used:</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span>psamm-model primarypairs --weights <span class="name variable">N</span><span class="operator">=</span>0.2,C<span class="operator">=</span>1,H<span class="operator">=</span>0,*<span class="operator">=</span>0.82 --exclude &#64;./exclude.tsv
</pre>
<p>In the case of a small model like the E. coli core model the results of <cite>primarypairs</cite> will likely not change
unless the weights are drastically altered. Changes could be seen in larger models though especially if the
models include many reactions related to non-carbon metabolism such as sulfur or nitrogen metabolism.</p>
</div>
<div class="section" id="report-element">
<h2>Report Element</h2>
<p>By default the <cite>primarypairs</cite> result is not filtered to show transfers of any specific element. In certain situations
it might be desirable to only get a subset of these results based on if the reactant/product pair transfers a target
element. To do this the option <cite>--report-element</cite> can be used. In many cases it might be desirable to only report
carbon transferring reactant/product pairs, to do this run the following on the E. coli model.</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span>psamm-model primarypairs --report-element C --exclude &#64;./exclude.tsv
</pre>
<p>If the predicted pairs are looked at for one of the mannitol pathway reactions, MANNIDEH, the following can be seen:</p>
<pre class="code shell literal-block">
MANNIDEH        manni<span class="operator">[</span>c<span class="operator">]</span>        fru_c<span class="operator">[</span>c<span class="operator">]</span>        C6H12O6
MANNIDEH        nad_c<span class="operator">[</span>c<span class="operator">]</span>        nadh_c<span class="operator">[</span>c<span class="operator">]</span>       C21H26N7O14P2
</pre>
<p>If this result is compared to the results without the <cite>--report-element C</cite> option it can be seen that when
there are additional transfers in this reaction but they only involve hydrogen.</p>
<pre class="code shell literal-block">
MANNIDEH        manni<span class="operator">[</span>c<span class="operator">]</span>        nadh_c<span class="operator">[</span>c<span class="operator">]</span>       H
MANNIDEH        manni<span class="operator">[</span>c<span class="operator">]</span>        h_c<span class="operator">[</span>c<span class="operator">]</span>  H
MANNIDEH        manni<span class="operator">[</span>c<span class="operator">]</span>        fru_c<span class="operator">[</span>c<span class="operator">]</span>        C6H12O6
MANNIDEH        nad_c<span class="operator">[</span>c<span class="operator">]</span>        nadh_c<span class="operator">[</span>c<span class="operator">]</span>       C21H26N7O14P2
</pre>
</div>
<div class="section" id="pair-prediction-methods">
<h2>Pair Prediction Methods</h2>
<p>Two reactant/product pair prediction algorithms are implemented in the <cite>PSAMM</cite> <tt class="docutils literal">primarypairs</tt> command.
The default algorithm is the <cite>FindPrimaryPairs</cite> algorithm. The other algorithm that is
implemented is the <cite>Mapmaker</cite> algorithm. These algorithm can be chosen through the <tt class="docutils literal"><span class="pre">--method</span></tt> argument.</p>
<pre class="code shell literal-block">
<span class="name variable">$ </span>psammm-model primarypairs --method fpp
or
<span class="name variable">$ </span>psamm-model primarypairs --method mapmaker
</pre>
</div>
</div>
<div class="section" id="visualizating-models-using-psamm">
<h1><a class="toc-backref" href="#id5">Visualizating Models using PSAMM</a></h1>
<p>The <tt class="docutils literal">vis</tt> function in <cite>PSAMM</cite> can be used to convert text based YAML models to graph
based representations of the metabolism. The graph based representation contains two sets
of nodes, with one set representing the metabolites in the model, and the other set
representing reactions. These nodes are connected through edges that are determined
based on element transfer patters predicted through using the <cite>FindPrimaryPairs</cite>
algorithm. The <tt class="docutils literal">vis</tt> function provides multiple options to customize the graph
representation of the metabolism including, customized node labels, showing <cite>FBA</cite>
results, and changing node colors.</p>
<div class="section" id="basic-visualization">
<h2>Basic visualization</h2>
<p>The basic <tt class="docutils literal">vis</tt> function can be run through the following command:</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span>psamm-model vis
</pre>
<p>The basic visualization relies on the <cite>FindPrimaryPairs</cite> algorithm to predict
element transfer in the metabolic network. This algorithm requires certain
reactions such as biosynthesis reactions and other artificial reactions to be
excluded from the algorithm's calculations in order to work efficiently. This
can be done through the <tt class="docutils literal"><span class="pre">--exclude</span></tt> option. By default the biomass reaction
that is specified in the <cite>model.yaml</cite> file will be excluded from the <cite>FPP</cite>
calculation but will still be shown on the final graph image. For more
information on on excluded reactions see <a href="#id1"><span class="problematic" id="id2">:ref:`exclude-fpp`</span></a>.</p>
<div class="system-message" id="id1">
<p class="system-message-title">System Message: ERROR/3 (<tt class="docutils">fpp.rst</tt>, line 266); <em><a href="#id2">backlink</a></em></p>
Unknown interpreted text role &quot;ref&quot;.</div>
<p>In this version of the E. coli core model the biomass reaction is defined in
the <cite>model.yaml</cite> file so that it will be excluded automatically.</p>
<p>When running the above command three files will be produced called 'reaction.dot',
'reactions.edges.tsv', 'reactions.nodes.tsv'.</p>
<p>The first file, 'reactions.dot', contains a text based representation of the
network graph in the 'dot' language. This graph language is used primarily by
the <cite>Graphviz</cite> program to generate network images. This graph format contains
information on the nodes and edges in the graph along with details related to
the size, colors, and shapes that will be used in the final graph image.</p>
<p>The 'reactions.nodes.tsv', and 'reactions.edges.tsv' files are tab separated
tables that contain the same information as the <cite>dot</cite> based graph does, but
in a more generic table based format that can be used with other graph
analysis and visualization software like <cite>Cytoscape</cite>.</p>
<p>The 'reactions.edges.tsv' file contains the information related to the
structure of the graph. Each line in this table represents one edge in
the graph and contains the source of that edge, destination of that edge,
and the direction of the edge (forward, back, or both). This file will
look like the following:</p>
<pre class="code shell literal-block">
<span class="name builtin">source       </span>target  dir
CS_3 cit_c<span class="operator">[</span>c<span class="operator">]</span>        forward
f6p_c<span class="operator">[</span>c<span class="operator">]</span>     Biomass_Ecoli_core_w_GAM_5      forward
ALCD2x_1     acald_c<span class="operator">[</span>c<span class="operator">]</span>      both
...
</pre>
<p>The 'reacitons.nodes.tsv' file will contain all of the information related
to what appears on the nodes in the graph. In this representation of the
graph there both reactions nodes and compound nodes. This file will look
like the following:</p>
<pre class="code shell literal-block">
id   compartment     fillcolor       shape   style   <span class="name builtin">type    </span>label
13dpg_c<span class="operator">[</span>c<span class="operator">]</span>   c       <span class="comment">#ffd8bf ellipse filled  cpd     13dpg_c[c]
</span>2pg_c<span class="operator">[</span>c<span class="operator">]</span>     c       <span class="comment">#ffd8bf ellipse filled  cpd     2pg_c[c]
</span>3pg_c<span class="operator">[</span>c<span class="operator">]</span>     c       <span class="comment">#ffd8bf ellipse filled  cpd     3pg_c[c]
</span>6pgc_c<span class="operator">[</span>c<span class="operator">]</span>    c       <span class="comment">#ffd8bf ellipse filled  cpd     6pgc_c[c]</span>
</pre>
</div>
<div class="section" id="generating-image-files-from-text-based-graphs">
<h2>Generating Image Files from text based Graphs</h2>
<p>Images can be generated from the 'reactions.dot' file by using the <cite>Graphviz</cite>
program. For the best representations of the metabolic network the <cite>dot</cite> layout
is recommended. The image file can be generated as a <cite>PDF</cite> file by using the
following <cite>Graphviz</cite> program command:</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span>dot -O Tpdf reactions.dot
</pre>
<p>This can also be done at the same time as the <cite>vis</cite> command step by adding the
<tt class="docutils literal"><span class="pre">--image</span></tt> option with an image format (pdf, svg, eps, etc.) to the command:</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span>psamm-model vis --Image pdf
</pre>
<p>These commands will generate an image file called 'reactions.dot.pdf'. This
image file is the representation of what was in the 'reactions.dot' file. This
graph will look like:</p>
<img alt="6_basic_visual.png" src="6_basic_visual.png" />
<p>In this default version of the graph there are two sets of nodes: circular orange
nodes representing metabolites and rectangular green nodes representing reactions.
These nodes are connected by edges that will show the direction of the reaction
in the original metabolic model. This graph also contains some reaction nodes that
have multiple reaction IDs. These represent reactions that contain common
reactant/product pairs such as ATP/ADP.</p>
<p>The rest of the tutorial will deal with how to alter the graph representation from
this default format to show different aspects of the metabolism and customize the look
and content of the graph. For these sections, the mannitol utilization pathway that has
been added and used in the previous tutorial sections will be used as an example.</p>
</div>
<div class="section" id="representing-different-element-flows">
<h2>Representing Different Element Flows</h2>
<p>By default the <tt class="docutils literal">vis</tt> command will produce a graph that shows the carbon transfers
in the metabolic network. In the <tt class="docutils literal">primarypairs</tt> tutorial section above the element
transfers in the <cite>ACKr</cite> reaction were examined to see how the <cite>FPP</cite> algorithm would
predict element transfer patterns. The <tt class="docutils literal">vis</tt> command will use the information in
these element transfer predictions to filter the graph that is made to only show
certain edges. In the case of the <cite>ACKr</cite> reaction, if the element being shown is
carbon then only the 'Acetate -&gt; Acetyl-Phosphate' and 'ATP -&gt; ADP' edges would
be shown in the final graph. The 'ATP -&gt; Acetyl-Phosphate' edge will not be shown
in the draft becuase that element transfer only contains phosphate and oxygen.</p>
<div class="system-message">
<p class="system-message-title">System Message: ERROR/3 (<tt class="docutils">fpp.rst</tt>, line 373)</p>
<p>Malformed table.
Text in column margin in table line 4.</p>
<pre class="literal-block">
=====================         ================
Reactant/Product Pair         Element Transfer
=====================         ================
Acetate -&gt; Acetyl-Phosphate   carbon backbone
ATP -&gt; ADP                    carbon backbone
ATP -&gt; Acetyl-Phosphate       phosphate group
Acetate -&gt; ADP                None
=====================         ================
</pre>
</div>
<p>This type of element filtering can provide different views of the metabolic
network by showing how metabolic pathways transfer different elements through
the reactions. The mannitol utilization pathway that was added to the model
contains a multiple step pathway that converts extracellular mannitol to
fructose 6-phosphate. This pathway also involves multiple phosphorylation
and dephosphorylation steps. The <tt class="docutils literal"><span class="pre">--element</span></tt> argument can be added to the
the <tt class="docutils literal">vis</tt> command to filter this pathway to show the transfer patterns
of the phosphorous in the pathway:</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-model<span class="operator">)</span> <span class="name variable">$ </span>psamm-model vis --element P --Image pdf
</pre>
<p>The resulting 'reactions.dot.pdf' file will contain the phosphorous transfer
network of the E. coli core model.</p>
<p>If the mannitol utilization pathway is examined more closely it will be seen
that the transfers are different from the original and involve transfers
between metabolites like phosphate (pi[c]) that are typically filtered out of
the carbon only graph.</p>
<img alt="element-transfers.png" src="element-transfers.png" />
</div>
<div class="section" id="showing-cellular-compartments">
<h2>Showing Cellular Compartments</h2>
<p>GEMs often contain some representation of cellular compartments. At the most
basic level this might just include an intracellular and extracellular compartment,
but in more complex models additional compartments such as the periplasm in bacteria
or the mitochondria in eukaryotes can be included to represent more complex metabolic
processes. The <tt class="docutils literal">vis</tt> command can show these compartments in the final image through
the use of the <tt class="docutils literal"><span class="pre">--compartment</span></tt> argument. If the compartment information is not
defined in the model.yaml file then the the command will attempt to
automatically detect the organization of the compartments by examining the reaction
equations in the model. This process cannot always accurately predict the compartment
organization though. To overcome this problem it is better to define the compartment
organization in the model.yaml file. This can be done by adding a new <tt class="docutils literal">compartments</tt>
sections into the model.yaml file like so:</p>
<pre class="code shell literal-block">
name: Ecoli_core_model
biomass: Biomass_Ecoli_core_w_GAM
default_flux_limit: 1000
extracellular: e
compartments:
- id: c
  adjacent_to: e
  name: Cytoplasm
- id: e
  adjacent_to: e
  name: Extracellular
....
</pre>
<p>The first thing to define is the extracellular compartment through the line
'extracellular: {id}'. This will tell <cite>PSAMM</cite> what the outermost compartment
is in the model. The next thing to define is the compartments section. In this
section each compartment can be defined with an id, a name, and a list of compartments
that are adjacent to that compartment.</p>
<p>Once this information is added to the model.yaml file the following command can
be used to generate an image that shows the compartment information of the model:</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span>psamm-model vis --compartment --Image
</pre>
<p>This command will produce an additional file called 'reactions_compartmentalized.dot'
and a new image file 'reactions_compartmentalized.dot.pdf' in addition to the
other files that are normally made. The 'reactions_compartmentalized.dot.pdf'
image will look like this:</p>
<img alt="13_compartmentalized.dot.png" src="13_compartmentalized.dot.png" />
<p>In this image there will be two compartments that are labeled with 'e' and 'c'.
The E. coli core model is relatively small, meaning that the compartment
organization is simple, but the <tt class="docutils literal">vis</tt> command can handle more complex
models as well. For example the following image was made using a toy model
to show a more complex compartments organization.</p>
<img alt="14_toy_compartment.png" src="14_toy_compartment.png" />
</div>
<div class="section" id="visualizing-reactions-and-pathways-of-interest">
<h2>Visualizing Reactions and Pathways of Interest</h2>
<p>In some situations it might be better to only visualize a subset of a larger
model so that smaller subsystems can be examined in more detail. This can
be done through the <tt class="docutils literal"><span class="pre">--subset</span></tt> argument. this argument will take an input
of a single column file where each line contains either a reaction ID or
a metabolite ID. The whole file can only contain reaction IDs or metabolite
IDs and cannot be a mix of both in the same subset file.</p>
<p>To use this command to make a subset of the reactions from the model that
are involved in the mannitol utilization pathway the following command can
be used:</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span>psamm-model vis --subset ../additional_files/subset_mannitol_pathway --Image pdf
</pre>
<p>The input file subset_mannitol_pathway looks like this:</p>
<pre class="code shell literal-block">
MANNIPTS
MANNI1PDEH
MADNNIDEH
MANNII1PPHOS
FRUKIN
</pre>
<p>This subset image will look like the following:</p>
<img alt="15_subset_mannitol_pathway.png" src="15_subset_mannitol_pathway.png" />
<p>In this image only the reactions directly listed in the subset file are included in
the final image.</p>
<p>The other option for using the subset argument is to provide a list of metabolite IDs.
This option will generate an image containing all of the reactions that the given
metabolites are involved in. For example the following subset file could be used to
generate a network image that contains all of the reactions involving pyruvate in
this model.</p>
<pre class="code shell literal-block">
pyr_c<span class="operator">[</span>c<span class="operator">]</span>
pyr_e<span class="operator">[</span>e<span class="operator">]</span>
</pre>
<p>To use this subset in to generate the pyruvate related subnetwork use the following
command:</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span>psamm-model vis --subset ../additional_files/subset_pyruvate_list --Image pdf
</pre>
<p>This will generate an image like the following that only shows the reactions that
contain the pyruvate metabolite:</p>
<img alt="16_subset_pyr_image.png" src="16_subset_pyr_image.png" />
</div>
<div class="section" id="highlighting-reactions-and-metabolites-in-the-network">
<h2>Highlighting reactions and Metabolites in the Network</h2>
<p>The <tt class="docutils literal"><span class="pre">--subset</span></tt> argument can be used to show only a specific part of the network.
When this is done the context of those reactions that are shown is often lost
and it can be hard to tell where that pathway fits withing the larger metabolism.
A different way to highlight a set of reactions without using the <tt class="docutils literal"><span class="pre">--subset</span></tt>
option is to change the color of a set of nodes through the <tt class="docutils literal"><span class="pre">--color</span></tt> argument.</p>
<p>This argument can be used to change the color of the reaction or metabolite nodes
on the final graph, making it easy to highlight certain pathways while still
maintaining the larger metabolic context. This <tt class="docutils literal"><span class="pre">--color</span></tt> argument will take a
two column file that contains reaction or metabolite IDs in the first column and
hex color codes in the second column. A color file that can be used to color
all of the mannitol utilization pathway reactions purple would look like the following:</p>
<pre class="code shell literal-block">
MANNIPTS <span class="comment">#d6c4f2
</span>MANNI1PDEH <span class="comment">#d6c4f2
</span>MANNIDEH <span class="comment">#d6c4f2
</span>MANNI1PPHOS <span class="comment">#d6c4f2
</span>FRUKIN <span class="comment">#d6c4f2</span>
</pre>
<p>To use this file to generate an image of the larger network with the mannitol
utilization pathway highlighted use the following command:</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span>vis --color ../additional_files/color_mannitol_pathway --Image pdf
</pre>
<p>The resulting image file should look like the following:</p>
<img alt="19_recolor_mannitol.png" src="19_recolor_mannitol.png" />
<p>Coloring of specific nodes like this can make it easy to locate or highlight
specific pathways especially in larger models.</p>
</div>
<div class="section" id="modify-node-labels-in-network-images">
<h2>Modify Node Labels in Network Images</h2>
<p>By default only the reaction IDs or metabolite IDs are shown on the nodes in the
final network images. These labels can be modified to include any additional
information defined in the compounds or reactions file through the use of the
<tt class="docutils literal"><span class="pre">--cpd-detail</span></tt> and <tt class="docutils literal"><span class="pre">--rxn-detail</span></tt> arguments. These arguments can be
followed by a space separated list of properties such as id, name,
equation, and formula which will be included on the node labels in the final image.
To apply these options to the model to show the ID, names, and formulas for the
metabolites and the ID and equation for reactions the following command can be used:</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span>psamm-model vis --subset ../additional_files/subset_mannitol_pathway --cpd-detail id name formula --rxn-detail id genes equation --Image pdf
</pre>
<p>The image generated will look like this:</p>
<img alt="17_detail.png" src="17_detail.png" />
<p>For these options if a detail is provided that is not a property of the reaction
or compound then that property will be skipped and not included on those nodes. For
example if the formula property is given to the <tt class="docutils literal"><span class="pre">-rxn-detail</span></tt> option then that
property will be skipped.</p>
</div>
<div class="section" id="showing-reaction-fluxes-on-pathway-images">
<h2>Showing Reaction Fluxes on Pathway Images</h2>
<p>Flux Balance analysis can be used to simulate reaction fluxes in a metabolic
model. These flux values can be used to modify the edges on pathway images to
provide a different way to look at and help interpret these flux results. This
can be done easily through the <tt class="docutils literal"><span class="pre">--fba</span></tt> option. When using this option a
l1min <tt class="docutils literal">fba</tt> will be run maximizing the biomass reaction of the model. The
reaction fluxes will then be used to adjust the width of the edges in the graph.
Thicker edges will represent larger flux values and reactions that carry no flux
will be shown as dashed lines. This can be done through the following command:</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span>psamm-model vis --fba
</pre>
<img alt="21_fba.png" src="21_fba.png" />
</div>
<div class="section" id="other-visualization-options">
<h2>Other Visualization Options</h2>
<div class="section" id="splitting-condensed-nodes">
<h3>Splitting Condensed Nodes</h3>
<p>By default the vis command will combined reaction nodes that have the same
reactant/product pairs. This is often seen with reactant/product pairs like
ATP/ADP and NADH/NADH. When these nodes are condensed they will not have the
any modifications through the <tt class="docutils literal"><span class="pre">--color</span></tt>, <tt class="docutils literal"><span class="pre">--rxn-detail</span></tt>, or <tt class="docutils literal"><span class="pre">--cpd-detail</span></tt>
arguments shown on the final image. To get these modifications to be shown on
the final image, these combined nodes can be split using the <tt class="docutils literal"><span class="pre">--split</span></tt> argument.
An example of this can be seen with the following command which will allow for the
the coloring of nodes that would normally be condensed:</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span>psamm-model vis --color ../additional_files/color_mannitol_pathway --split --Image pdf
</pre>
<img alt="20_recolor_slit.png" src="20_recolor_slit.png" />
</div>
<div class="section" id="removing-specific-reactant-product-pairs">
<h3>Removing Specific Reactant Product Pairs</h3>
<p>Larger scale models may have some reactant/product pairs that occur many times
in different reactions. These often involve currency metabolites like ATP, ADP, NAD,
and NADH. Due to the number of times that these pairs occur across the network they
can cause some parts of the graph to look messy. While making the condensed reaction
nodes helps with this problem, there may be cases where it would be better to
hide these edges in the final result. To do this the <tt class="docutils literal"><span class="pre">--hide-edges</span></tt> argument can
be used. This argument takes a two column file where each row contains reaction IDs
where the edges between them will be hidden.</p>
<div class="note">
<p class="first admonition-title">Note</p>
<p class="last">Needs better example. perhaps with a combined image like was done with the
element transfers example.</p>
</div>
</div>
<div class="section" id="adjusting-image-size">
<h3>Adjusting Image Size</h3>
<p>The size of the final image file that is produced through the <tt class="docutils literal">vis</tt> command can
be adjusted through the <tt class="docutils literal"><span class="pre">--image-size</span></tt> argument. This argument will take the width
and height (in inches) separated by a comma. The following command can be used to
generate an image that is 8.5 inches x 11 inches:</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span>psamm-model vis --Image pdf --image-size 8.5,11
</pre>
</div>
<div class="section" id="changing-pair-prediction-methods">
<h3>Changing Pair Prediction Methods</h3>
<p>The <tt class="docutils literal">vis</tt> function in <cite>PSAMM</cite> will use the <cite>FPP</cite> algorithm by default. This
can be changed to either allow for a file with a precalculated pair prediction
or for no pair prediction to be used instead of <cite>FPP</cite>. This can be done through the
<tt class="docutils literal"><span class="pre">--method</span></tt> option.</p>
<p>To use this argument with a pair prediction file first a file must be prepared that
contains the pair prediction results. This file would be in the format that the <cite>PSAMM</cite>
<tt class="docutils literal">primarypairs</tt> function produces, with four columns that contain a reaction ID, substrate ID,
product ID, and elements transferred. The file would look like the following:</p>
<pre class="code shell literal-block">
FBA fdp_c<span class="operator">[</span>c<span class="operator">]</span>   g3p_c<span class="operator">[</span>c<span class="operator">]</span> C3O6P
PPS   h2o_c<span class="operator">[</span>c<span class="operator">]</span> pi_c<span class="operator">[</span>c<span class="operator">]</span>  O
RPI   r5p_c<span class="operator">[</span>c<span class="operator">]</span> ru5p_D_c<span class="operator">[</span>c<span class="operator">]</span> C5O8P
....
</pre>
<p>This file can be supplied to the command through the <tt class="docutils literal"><span class="pre">--method</span></tt> option. For
example an image could be generated using a precalculated pair prediction using
the <cite>Mapmapker</cite> algorithm through the following command:</p>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span>psamm-model vis --method ../additional_files/mapmaker_prediction
</pre>
<p>In addition to the different prediction methods it is possible to generate a graph
that does not use any pair prediction method. this graph will just generate edges
based on the reaction equations themselves. When using this option there will tend
to be many more connections in the network image especially to metabolites like ATP,
H2O, and H+. This option can be used through the following command:</p>
<div class="note">
<p class="first admonition-title">Note</p>
<p class="last">The <tt class="docutils literal"><span class="pre">--method</span> <span class="pre">no-fpp</span></tt> and <tt class="docutils literal"><span class="pre">--split</span></tt> arguments cannot be used together. The
<tt class="docutils literal"><span class="pre">--split</span></tt> option will only act on condensed nodes in the graph, these nodes
can only be generated if some pair prediction is provided.</p>
</div>
<pre class="code shell literal-block">
<span class="operator">(</span>psamm-env<span class="operator">)</span> <span class="name variable">$ </span>psamm-model vis --method no-fpp
</pre>
<img alt="10_no_fpp.png" src="10_no_fpp.png" />
</div>
</div>
</div>
</div>
</body>
</html>
